crates/subspace-farmer/src/pieces_verification.rs

```diff
@@ -26,25 +29,36 @@ pub enum PiecesVerificationError {
 /// Verifies pieces against the blockchain.
 pub async fn verify_pieces_at_blockchain<RC: RpcClient>(
     verification_client: &RC,
+    farmer_protocol_info: FarmerProtocolInfo,
     piece_indexes: &[PieceIndex],
     pieces: &FlatPieces,
 ) -> Result<(), PiecesVerificationError> {
-    if piece_indexes.len() != (pieces.len() / PIECE_SIZE) {
+    if piece_indexes.len() != pieces.count() {
         return Err(PiecesVerificationError::InvalidRawData);
     }
-    for (index, piece) in pieces.as_pieces().enumerate() {
-        let piece_index = piece_indexes[index];
 
-        let segment_index: u64 = piece_index / MERKLE_NUM_LEAVES as u64;
-        let position: u64 = piece_index % MERKLE_NUM_LEAVES as u64;
+    let records_per_segment =
+        farmer_protocol_info.recorded_history_segment_size as usize / PIECE_SIZE;
+    if farmer_protocol_info.record_size.is_zero() || records_per_segment.is_zero() {
```

Comment:
It would be preferable to change `record_size` and `recorded_history_segment_size` to `NonZeroU32`, then you don't ever need to check it. At least in `FarmerProtocolInfo` :slightly_smiling_face: 

It is always better to make invalid invariants non-representable in the first place. Though I don't think these two were ever zero anywhere in our codebase anyway. And segment size was always bigger than piece size.